lib/transformer.js

```diff
@@ -0,0 +1,228 @@
+
+    for (const file of parsed) {
+      for (const instruction of file.idl) {
+        let oldMembers;
+        let extAttrs;
+        switch (instruction.type) {
+          case "interface":
+            if (!instruction.partial) {
+              break;
+            }
+
+            if (this.options.suppressErrors && !interfaces[instruction.name]) {
+              break;
+            }
+            oldMembers = interfaces[instruction.name].idl.members;
+            oldMembers.push.apply(oldMembers, instruction.members);
+            extAttrs = interfaces[instruction.name].idl.extAttrs;
+            extAttrs.push.apply(extAttrs, instruction.extAttrs);
+            break;
+          case "dictionary":
+            if (!instruction.partial) {
+              break;
+            }
+            if (this.options.suppressErrors && !dictionaries[instruction.name]) {
+              break;
+            }
+            oldMembers = dictionaries[instruction.name].idl.members;
+            oldMembers.push.apply(oldMembers, instruction.members);
+            extAttrs = dictionaries[instruction.name].idl.extAttrs;
+            extAttrs.push.apply(extAttrs, instruction.extAttrs);
+            break;
+          case "implements":
+            if (this.options.suppressErrors && !interfaces[instruction.target]) {
+              break;
+            }
+            interfaces[instruction.target].implements(instruction.implements);
+            break;
+        }
+      }
+    }
+  }
+
+  *_writeFiles(outputDir) {
+    let utilsText = fs.readFileSync(__dirname + "/output/utils.js");
+    fs.writeFileSync(this.options.utilPath, utilsText);
+
+    let interfaceStub = fs.readFileSync(__dirname + "/output/interfaceStub.js");
+    let keys = Object.keys(this._interfaces).concat(Object.keys(this._dictionaries));
+    for (let key of keys) {
+      if (this._interfaces[key]) {
+        fs.writeFileSync(path.join(outputDir, this._interfaces[key].name + ".js"), interfaceStub);
+      } else {
+        fs.writeFileSync(path.join(outputDir, this._dictionaries[key].name + ".js"), "");
+      }
+    }
+
+    keys = Object.keys(this._interfaces);
+    for (let i = 0; i < keys.length; ++i) {
+      const obj = this._interfaces[keys[i]];
+      let source = obj.toString();
+
+      const implDir = path.relative(outputDir, obj.opts.implDir).replace(/\\/g, "/"); // fix windows file paths
+      let implFile = implDir + "/" + obj.name + this.options.implSuffix;
+      if (implFile[0] !== ".") {
+        implFile = "./" + implFile;
+      }
+
+      let relativeUtils = path.relative(outputDir, this.options.utilPath).replace(/\\/g, '/');
+      if (relativeUtils[0] !== ".") {
+        relativeUtils = "./" + relativeUtils;
+      }
+
+      source = `"use strict";
+
+const conversions = require("webidl-conversions");
+const utils = require("${relativeUtils}");\n${source}
+const Impl = require("${implFile}.js");\n`;
+
+      fs.writeFileSync(path.join(outputDir, obj.name + ".js"), source);
```

Comment:
Don't do sync file writes if you're trying to promise an async interface... same below on line 212